// @awo00/smb2/dist/protocol/Packet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.protocolIdStructureField = void 0;
const protocolIds = __importStar(require("./protocolIds"));
const structureUtil = __importStar(require("./structureUtil"));
exports.protocolIdStructureField = {
    type: String,
    encoding: "hex",
    size: 4,
    defaultValue: protocolIds.smb2
};
class Packet {
    static parseProtocolId(buffer) {
        return structureUtil.parseValue(buffer, exports.protocolIdStructureField);
    }
    static getChunks(buffer) {
        const chunks = [];
        while (buffer.length > 4) {
            const netBiosType = buffer.readUInt8(0);
            if (netBiosType !== 0x00)
                throw new Error("no_net_bios_message");
            const packetLength = buffer.readUInt32BE(0);
            if (packetLength > buffer.length - 4)
                break;
            buffer = buffer.slice(4);
            chunks.push(buffer.slice(0, packetLength));
            buffer = buffer.slice(packetLength);
        }
        return {
            chunks,
            restChunk: buffer
        };
    }
}
exports.default = Packet;
